File: 10-edge/04-quantization-and-determinism/golang/parallel_sum.go
// Layer 10 - Topic 4: the batch-invariance finding, on a CPU, with no
// kernels and no GPU. (Go)
//
// What this demonstrates
//
//	Thinking Machines Lab's Defeating Nondeterminism in LLM Inference
//	(September 2025) traced temperature-0 nondeterminism to kernels that
//	are not BATCH-INVARIANT: a reduction split differently across a
//	different batch shape sums the same numbers in a different order,
//	floating-point addition is not associative, and the last bits differ.
//	Occasionally those bits cross an argmax boundary between two close
//	logits and the sequences diverge completely.
//
//	This file is that mechanism with everything else removed. Ten million
//	float32 values, one slice, summed by W goroutines and combined. W is
//	the only thing that changes. The Go spec is strict about float
//	semantics -- no reassociation, no surprise FMA -- so what you are
//	seeing is purely partition order.
//
// What to look for
//
//   - `distinct sums` across the W values. More than one means the
//     result of a pure function of a fixed input depended on how the
//     work was divided.
//
//   - Bit patterns (%b prints the exact mantissa and exponent). Two sums
//     that print the same in %f can differ here, which is why decimal
//     formatting is not a way to check determinism.
//
//   - `same W, repeated` at the bottom: each W is DETERMINISTIC on its
//     own. This is not a race. Run it a hundred times and W=8 always
//     gives the same answer -- it is just a different answer from W=4.
//
//   - The relative spread against float32's epsilon. It is tiny, and
//     tiny is exactly the size of an argmax flip between two close
//     logits.
//
//     Then the sentence to carry out of this topic: your result depended
//     on how the work was partitioned, and the partitioning depended on
//     load. Go find that bug in something you already own.
//
// No dependencies. Runs with no arguments (-workers overrides the default
// list):
//
//	cd golang && go run parallel_sum.go
//	cd golang && go run parallel_sum.go -workers 1,3,7,64
package main

import (
	"flag"
	"fmt"
	"math"
	"math/rand"
	"sort"
	"strconv"
	"strings"
	"sync"
)

const (
	n    = 10_000_000
	seed = 20260818
)

// sumPartitioned splits data across w goroutines, sums each chunk in
// float32, and adds the partials in order. Deterministic for a given w.
func sumPartitioned(data []float32, w int) float32 {
	if w <= 1 {
		var s float32
		for _, v := range data {
			s += v
		}
		return s
	}
	partials := make([]float32, w)
	chunk := (len(data) + w - 1) / w
	var wg sync.WaitGroup
	for i := 0; i < w; i++ {
		lo := i * chunk
		if lo >= len(data) {
			break
		}
		hi := lo + chunk
		if hi > len(data) {
			hi = len(data)
		}
		wg.Add(1)
		go func(i, lo, hi int) {
			defer wg.Done()
			var s float32
			for _, v := range data[lo:hi] {
				s += v
			}
			partials[i] = s
		}(i, lo, hi)
	}
	wg.Wait()
	// Combining in index order keeps this deterministic per w: the only
	// variable in the whole program is how many pieces there were.
	var total float32
	for _, p := range partials {
		total += p
	}
	return total
}

func main() {
	workersFlag := flag.String("workers", "1,2,4,8,16,32,64",
		"comma-separated goroutine counts")
	flag.Parse()

	var workers []int
	for _, f := range strings.Split(*workersFlag, ",") {
		v, err := strconv.Atoi(strings.TrimSpace(f))
		if err != nil {
			panic(err)
		}
		workers = append(workers, v)
	}

	rng := rand.New(rand.NewSource(seed))
	data := make([]float32, n)
	for i := range data {
		// Values around 1.0 so the running sum grows to ~1e7 while each
		// addend stays ~1 -- the regime where float32 starts losing the
		// low bits of every addition, which is the regime a real reduction
		// over activations lives in.
		data[i] = float32(rng.Float64() + 0.5)
	}

	// The reference: the same numbers summed in float64, which has enough
	// bits that partition order does not reach the printed digits.
	var exact float64
	for _, v := range data {
		exact += float64(v)
	}

	fmt.Println("Partition-order nondeterminism -- Go, float32, no GPU involved")
	fmt.Printf("  %d values ~U(0.5, 1.5), seed %d\n", n, seed)
	fmt.Printf("  float64 reference sum: %.6f\n\n", exact)

	fmt.Printf("  %8s %20s %26s %14s\n", "workers", "sum (float32)", "exact bits", "rel err")
	fmt.Println("  " + strings.Repeat("-", 72))
	seen := map[float32]bool{}
	var values []float64
	for _, w := range workers {
		s := sumPartitioned(data, w)
		seen[s] = true
		values = append(values, float64(s))
		fmt.Printf("  %8d %20.6f %26b %14.3e\n", w, s, s,
			math.Abs(float64(s)-exact)/exact)
	}

	fmt.Printf("\n  distinct sums across %d partitionings: %d\n", len(workers), len(seen))
	sort.Float64s(values)
	spread := (values[len(values)-1] - values[0]) / exact
	fmt.Printf("  relative spread: %.3e   (float32 epsilon: %.3e)\n",
		spread, math.Nextafter32(1, 2)-1)

	fmt.Println("\n  same W, repeated -- this is not a race:")
	for _, w := range []int{4, 8} {
		a := sumPartitioned(data, w)
		b := sumPartitioned(data, w)
		c := sumPartitioned(data, w)
		fmt.Printf("    W=%-3d %b  %b  %b   identical: %v\n", w, a, b, c,
			a == b && b == c)
	}

	fmt.Println("\n  Each partitioning is perfectly reproducible on its own. What is")
	fmt.Println("  not reproducible is WHICH partitioning you get, because on a real")
	fmt.Println("  inference server that is chosen by the batch shape, and the batch")
	fmt.Println("  shape is chosen by other people's traffic. That is the whole of")
	fmt.Println("  the temperature-0 nondeterminism finding, with the GPU removed.")
}

File: 10-edge/04-quantization-and-determinism/java/ParallelSum.java
// Layer 10 - Topic 4: the batch-invariance finding, on a CPU. (Java)
//
// What this demonstrates
//     The same mechanism as golang/parallel_sum.go, through Java's own
//     parallel reduction. Ten million values, summed three ways:
//
//       sequential            DoubleStream.sum(), one thread
//       parallel              DoubleStream.parallel().sum(), where the
//                             result depends on how the SPLITERATOR chose
//                             to divide the work -- a decision made by the
//                             framework, based on the common ForkJoinPool
//                             and the input size
//       explicit W-way        the same partitioning written by hand, for
//                             W = 1..64, so you can see the answer move
//                             with W rather than with the weather
//
//     Note that DoubleStream.sum() is not a naive loop: it carries a
//     compensation term (Kahan summation), which is why its sequential
//     answer is closer to exact than a hand-written loop. That is worth
//     knowing before you conclude anything from a comparison against one.
//
// Worth knowing the history, because it is precisely this topic's journey:
//     `strictfp` used to be a keyword you needed for reproducible results,
//     because an x87 FPU could keep intermediates at 80 bits and the answer
//     depended on the host. JEP 306 (Java 17) made all floating-point
//     strict by default. So Java moved from "your result depends on the
//     hardware" to "your result depends only on your partitioning" -- which
//     is the exact problem this topic is about, and the reason the
//     remaining nondeterminism is now entirely yours.
//
// What to look for
//     - `distinct` across the explicit partitionings. More than one means
//       a pure function of a fixed input gave different answers depending
//       on how it was divided.
//     - Double.toHexString, not %f. Two sums that print identically in
//       decimal can differ in the bits that matter.
//     - The float32 section. Repeat the same experiment at float precision
//       and the spread grows by orders of magnitude -- which is the regime
//       an inference kernel actually runs in.
//
// No dependencies. Runs with no arguments:
//     cd java && javac ParallelSum.java -d /tmp/javabuild \
//       && java -cp /tmp/javabuild ParallelSum

import java.util.Arrays;
import java.util.HashSet;
import java.util.Random;
import java.util.Set;
import java.util.concurrent.ForkJoinPool;
import java.util.stream.DoubleStream;

public class ParallelSum {
    static final int N = 10_000_000;
    static final long SEED = 20260818L;
    static final int[] WORKERS = {1, 2, 4, 8, 16, 32, 64};

    /** Sum in `w` chunks, each chunk summed in order, partials added in order. */
    static double sumPartitioned(double[] data, int w) {
        if (w <= 1) {
            double s = 0;
            for (double v : data) s += v;
            return s;
        }
        int chunk = (data.length + w - 1) / w;
        double[] partials = new double[w];
        ForkJoinPool.commonPool().submit(() ->
            java.util.stream.IntStream.range(0, w).parallel().forEach(i -> {
                int lo = i * chunk;
                int hi = Math.min(data.length, lo + chunk);
                double s = 0;
                for (int j = lo; j < hi; j++) s += data[j];
                partials[i] = s;
            })).join();
        double total = 0;
        for (double p : partials) total += p;   // in index order: deterministic per w
        return total;
    }

    static float sumPartitionedFloat(float[] data, int w) {
        if (w <= 1) {
            float s = 0;
            for (float v : data) s += v;
            return s;
        }
        int chunk = (data.length + w - 1) / w;
        float[] partials = new float[w];
        ForkJoinPool.commonPool().submit(() ->
            java.util.stream.IntStream.range(0, w).parallel().forEach(i -> {
                int lo = i * chunk;
                int hi = Math.min(data.length, lo + chunk);
                float s = 0;
                for (int j = lo; j < hi; j++) s += data[j];
                partials[i] = s;
            })).join();
        float total = 0;
        for (float p : partials) total += p;
        return total;
    }

    public static void main(String[] args) {
        Random rng = new Random(SEED);
        double[] data = new double[N];
        float[] dataF = new float[N];
        for (int i = 0; i < N; i++) {
            double v = rng.nextDouble() + 0.5;   // ~U(0.5, 1.5)
            data[i] = v;
            dataF[i] = (float) v;
        }

        System.out.println("Partition-order nondeterminism -- Java "
                + Runtime.version().feature() + ", no GPU involved");
        System.out.printf("  %d values ~U(0.5, 1.5), seed %d%n", N, SEED);
        System.out.printf("  common ForkJoinPool parallelism: %d%n%n",
                ForkJoinPool.getCommonPoolParallelism());

        double seq = DoubleStream.of(data).sum();
        double par = DoubleStream.of(data).parallel().sum();
        System.out.println("  DoubleStream (Kahan-compensated inside the JDK):");
        System.out.printf("    sequential          %s  %.9f%n", Double.toHexString(seq), seq);
        System.out.printf("    parallel()          %s  %.9f%n", Double.toHexString(par), par);
        System.out.printf("    identical bits: %s%n", seq == par ? "yes" : "NO");
        System.out.println("    (the spliterator decided the split; you did not, and the");
        System.out.println("     decision depends on input size and pool parallelism)");

        System.out.println("\n  Explicit W-way partitioning, double:");
        System.out.printf("    %8s %26s %20s %14s%n", "workers", "bits", "sum", "rel err");
        Set<Double> seen = new HashSet<>();
        double[] values = new double[WORKERS.length];
        for (int i = 0; i < WORKERS.length; i++) {
            double s = sumPartitioned(data, WORKERS[i]);
            seen.add(s);
            values[i] = s;
            System.out.printf("    %8d %26s %20.6f %14.3e%n", WORKERS[i],
                    Double.toHexString(s), s, Math.abs(s - seq) / seq);
        }
        Arrays.sort(values);
        System.out.printf("    distinct: %d of %d    relative spread: %.3e   "
                + "(double ulp near 1: %.3e)%n", seen.size(), WORKERS.length,
                (values[values.length - 1] - values[0]) / seq, Math.ulp(1.0));

        System.out.println("\n  Same experiment at float precision -- the regime a kernel");
        System.out.println("  actually runs in:");
        System.out.printf("    %8s %20s %20s %14s%n", "workers", "bits", "sum", "rel err");
        Set<Float> seenF = new HashSet<>();
        float[] valuesF = new float[WORKERS.length];
        for (int i = 0; i < WORKERS.length; i++) {
            float s = sumPartitionedFloat(dataF, WORKERS[i]);
            seenF.add(s);
            valuesF[i] = s;
            System.out.printf("    %8d %20s %20.6f %14.3e%n", WORKERS[i],
                    Float.toHexString(s), s, Math.abs(s - seq) / seq);
        }
        Arrays.sort(valuesF);
        System.out.printf("    distinct: %d of %d    relative spread: %.3e   "
                + "(float ulp near 1: %.3e)%n", seenF.size(), WORKERS.length,
                (valuesF[valuesF.length - 1] - valuesF[0]) / seq, Math.ulp(1.0f));

        System.out.println("\n  same W, repeated -- this is not a race:");
        for (int w : new int[] {4, 8}) {
            float a = sumPartitionedFloat(dataF, w);
            float b = sumPartitionedFloat(dataF, w);
            System.out.printf("    W=%-3d %s  %s   identical: %s%n", w,
                    Float.toHexString(a), Float.toHexString(b), a == b);
        }

        System.out.println("\n  Each partitioning is perfectly reproducible on its own. What");
        System.out.println("  is not reproducible is WHICH partitioning you get -- on an");
        System.out.println("  inference server that is decided by the batch shape, and the");
        System.out.println("  batch shape is decided by other people's traffic.");
        System.exit(0);
    }
}
